### ISSUE TITLE
LightningCLI + --trainer.gradient_clip_val XX doesn't work?

### ISSUE THREAD

**Issue by profPlum** [Lightning-AI/pytorch-lightning#18669]
### Bug description

When using LightningCLI + `--trainer.gradient_clip_val 0.25 + --trainer.gradient_clip_algorithm=value` tensorboard logs still show gradient infinity_norms > 0.25 (when using gradient norm logging). Infinity norm is the maximum value in the gradient, so it should be capped at the value set by --trainer.gradient_clip_val 0.25 (when algorithm=value also). But tensorboard logs indicate this isn't happening.

### What version are you seeing the problem on?

v1.9

### How to reproduce the bug

```python
LightningCLI + `--trainer.gradient_clip_val 0.25 + --trainer.gradient_clip_algorithm=value`


# Track grad norm, instructions from here: 
# https://github.com/Lightning-AI/lightning/pull/16745
def on_before_optimizer_step(self, optimizer, *args):
    self.log_dict(grad_norm(self, norm_type='inf')) # inspect (unscaled) gradients here
```

### Error messages and logs

<img width="251" alt="image" src="https://github.com/Lightning-AI/lightning/assets/4739054/5f0377d8-1632-40a3-a5d9-2f99f13ef03d">
<img width="330" alt="image" src="https://github.com/Lightning-AI/lightning/assets/4739054/2a239237-db74-4036-9920-3bd3e5c4c597">

### Environment

[env_details.txt](https://github.com/Lightning-AI/lightning/files/12773141/env_details.txt)

**Comment by profPlum** [comment]
Why is it failing to upload images & system details?

**Comment by awaelchli** [comment]
Hey @profPlum 
The documentation of `on_before_optimizer_step` states:

> If clipping gradients, the gradients will not have been clipped yet.

To log the gradients after clipping, I think you could do it by overriding `configure_gradient_clipping` and calling `super().configure_gradient_clipping()` first, or by overriding `optimizer_step()` (also call super first).
